File: Rally/src/com/fenway/rally/Rally.java
package com.fenway.rally;
import java.io.BufferedWriter;
import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

import com.google.gson.Gson;
import com.google.gson.JsonElement;

import com.rallydev.rest.*;
import com.rallydev.rest.request.QueryRequest;
import com.rallydev.rest.response.QueryResponse;
import com.rallydev.rest.util.Fetch;
import com.rallydev.rest.util.QueryFilter;

public class Rally {
	
	//Google GSON - Java serialization/deserialization library to convert Java Objects into JSON and back
	private static Gson gson = new Gson();
	
	/**
	 * Main method
	 * @param args
	 * @throws URISyntaxException
	 * @throws IOException
	 */
	public static void main(String[] args) throws URISyntaxException, IOException {
		//Check to see if a config file was specified - if not, exit out with a message
		if (args == null || args.length == 0 || args[0] == null || args[0].isEmpty()) {
			System.out.println("Please specify the associated config file in JSON format");
			int blah = 0;
			switch (blah) {
				case 0:
					System.out.print(blah);
				case 1:
					System.out.print(blah++);
				case 2:
					System.out.print(blah*2);
					break;
				default:
					System.out.print("Default");
			}
			return;
		}
		//Set the configuration file as what was specified in the command line
		File configFile = new File(args[0]);	
		
		//Retrieve configuration from specified JSON and set it to the configuration object
		StringBuilder configSB = new StringBuilder();
		try (Scanner scanner = new Scanner(configFile)) {
			while (scanner.hasNext()){
				configSB.append(scanner.nextLine());
			}
		}
		Config config = gson.fromJson(configSB.toString(), Config.class);
		
        //Create and configure a new instance of RallyRestApi
        RallyRestApi restApi = new RallyRestApi(new URI("https://rally1.rallydev.com"), config.getApiKey());
        restApi.setApplicationName("Rally");
        
        //Instantiate what will eventually become our JSON object
        Metric metric = new Metric();
        metric.setTeam_id(config.getTeamID());
        
        //Get the list of users to pull tasks (and thus stories) for
    	List<String> userList = config.getUserList();
        
        //Create a map of sprints
        Map<String,Iteration> sprintMap = new HashMap<String,Iteration>();
        try {
           	//Retrieve the previous and current sprint
        	sprintMap = getSprints(restApi, config.getProjectID());
        	
        	//Get the tasks (and consequentially, the stories) for each sprint
        	for (String key : sprintMap.keySet()) {
        		for (String user : userList) {
        			sprintMap.get(key).setStoryMap(getStories(restApi, sprintMap.get(key), user));
        		}
        		
        		//Now need to populate the actual story points per user based on hours worked
        		sprintMap.get(key).setStoryMap(distributePoints(sprintMap.get(key)));
        		
        		//Build out object which will become our JSON object
        		Sprint currSprint = new Sprint();
        		currSprint.setSprint_id(key);
        		//Set the start and end dates of the sprint using just the date component
        		currSprint.setSprint_start_date(sprintMap.get(key).getStartDate().substring(0, sprintMap.get(key).getStartDate().indexOf('T')));
        		currSprint.setSprint_end_date(sprintMap.get(key).getEndDate().substring(0, sprintMap.get(key).getEndDate().indexOf('T')));
        		//Get the list of stories for the sprint
        		List<Story> storyList = new ArrayList<Story>(sprintMap.get(key).getStoryMap().values());
        		currSprint.setStories(storyList);
        		//Add the sprint to the metrics object used to create the JSON output file
        		metric.addSprint(currSprint);
        	}
        	
        	//Convert the metric object to JSON
        	String metricJson = gson.toJson(metric);
        	
        	//Come up with filename
        	StringBuilder pathFileName = new StringBuilder();
        	pathFileName.append("c:/RallyStories/");
        	pathFileName.append(LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss")));
        	pathFileName.append("_");
        	pathFileName.append(configFile.getName());
        	
        	//Write out JSON to file
        	Path path = Paths.get(pathFileName.toString());
        	try (BufferedWriter writer = Files.newBufferedWriter(path)) 
        	{
        	    writer.write(metricJson);
        	    System.out.println("Successfully created JSON extract of metrics\n");
        	} catch (Exception e) {
        		System.out.println("Failed to create JSON extract of metrics - "+e.getMessage());
        	}
        	
        	//Also want to generate a CSV for easy pull-up of summary data per user per sprint
        	//Cheating with the filename - getting rid of the .json and replacing it with _sprints.csv
        	pathFileName.replace(pathFileName.length()-5, pathFileName.length(), "_sprints.csv");
        	path = Paths.get(pathFileName.toString());
        	try (BufferedWriter writer = Files.newBufferedWriter(path)) 
        	{
        		StringBuilder totalUserStoryPoints = new StringBuilder();
        		
        		for (Sprint sprint : metric.getSprint()) {
        			System.out.println("\n"+sprint.getSprint_id()+":");
        			Map<String,Double> users = new HashMap<String,Double>();
        			for (String u : config.getUserList()) {
        				users.put(u, 0.0);
        			}
        			//Loop through all the stories to add up the total story points per user in this sprint
        			for (Story story : sprint.getStories()) {
        				for (AssignedTo owner : story.getAssigned_to()) {
        					if (owner.getUser_points() != null) {
           						users.put(owner.getUser(), users.get(owner.getUser())+owner.getUser_points());
        					}
        				}
        			}
        			//Create output string for each user and their associated story hours
        			for (String user : users.keySet()) {
        				totalUserStoryPoints.append(sprint.getSprint_id()+","+user+","+users.get(user)+"\n");
        				System.out.println(user+" - "+users.get(user)+" points");
        			}
        			
        		}
        		writer.write(totalUserStoryPoints.toString());
        		System.out.println("Successfully created CSV of story points by user");
        	} catch (Exception e) {
        		System.out.println("Failed to create CSV extract of story points by user - "+e.getMessage());
        	}
        	
        } finally {
        	//Release all resources
            restApi.close();
        }
	}

	/**
	 * Retrieves the current and previous sprints in Rally
	 * @param restApi
	 * @return Map<String,Iteration>
	 * @throws IOException
	 */
	private static Map<String,Iteration> getSprints(RallyRestApi restApi, String projectID) throws IOException {
		//Set the begin and end dates for the sprints to pull back
		LocalDateTime startDate = LocalDateTime.now();
		LocalDateTime endDate = LocalDateTime.now();
		startDate = startDate.minusDays(28L);
		//startDate = startDate.minusDays(168L); //Used to get the last 6 sprints
		endDate = endDate.plusDays(14L);
		String startDateISO = startDate.format(DateTimeFormatter.ISO_DATE);
		String endDateISO = endDate.format(DateTimeFormatter.ISO_DATE);
		System.out.println("Retrieving last 2 sprints...");
		QueryRequest iterations = new QueryRequest("iteration");
		iterations.setFetch(new Fetch("Name", "Notes", "StartDate", "EndDate"));
		QueryFilter firstFilter = new QueryFilter("Project.ObjectUUID", "=", projectID);
		QueryFilter secondFilter = new QueryFilter("StartDate", ">=", startDateISO);
		QueryFilter thirdFilter = new QueryFilter("EndDate", "<=", endDateISO);
		iterations.setQueryFilter(firstFilter.and(secondFilter).and(thirdFilter));
		QueryResponse queryResponse = restApi.query(iterations);
		Map<String,Iteration> sprintMap = new HashMap<String,Iteration>();
		for (JsonElement result : queryResponse.getResults()) {
            Iteration sprint = gson.fromJson(result, Iteration.class);
            sprintMap.put(sprint.getName(), sprint);
        }
		return sprintMap;
	}
	
	/**
	 * Takes the restApi connection, the current Sprint, and the user to pull from and returns back a map of stories for that sprint and user
	 * @param restApi
	 * @param currSprint
	 * @param user
	 * @return Map<String,Story>
	 * @throws IOException
	 */
	private static Map<String,Story> getStories(RallyRestApi restApi, Iteration currSprint, String user) throws IOException{
		Map<String,Story> storyMap = (currSprint.getStoryMap() == null) ? new HashMap<String,Story>() : currSprint.getStoryMap();
		System.out.println(String.format("\n\nRetrieving stories for %s for %s...",user,currSprint.getName()));
		QueryRequest tasks = new QueryRequest("task");
		tasks.setFetch(new Fetch("WorkProduct", "Owner", "Project", "Release", "Iteration", "Name", "State", "PlanEstimate", "Estimate", "Actuals"));
		QueryFilter firstFilter = new QueryFilter("Iteration.Name", "=", currSprint.getName());
		QueryFilter secondFilter = new QueryFilter("Owner.Name", "=", user);
		tasks.setQueryFilter(firstFilter.and(secondFilter));
		QueryResponse queryResponse = restApi.query(tasks);
		System.out.println(String.format("\nTotal tasks: %d", queryResponse.getTotalResultCount()));
		for (JsonElement result : queryResponse.getResults()) {
			Task task = gson.fromJson(result,  Task.class);
			String story_id = task.getWorkProduct().getFormattedID();
			
			//Get the story associated with the task
			Story newStory;
			if (storyMap.containsKey(story_id)){
				//This story already exist in the story map, so update it
				newStory = storyMap.get(story_id);
			} else {
				//This is a new story, so create it
				newStory = createStory(restApi, story_id);
			}
			if (newStory != null) {
				AssignedTo newUser = new AssignedTo();
				
				//Set the owner properties for the story
				newUser.setUser(user);
				newUser.setUser_points(task.getActuals());
				
				//Set the story properties
				newStory.addAssigned_to(newUser);
				newStory.setStory_points(task.getWorkProduct().getPlanEstimate());
				
				//Save the new/updated story to the sprint's story map
				storyMap.put(story_id, newStory);
			}
			
        }
		
		return storyMap;
	}
	
	/**
	 * Creates a new story and assigns the story_id to it if the story is in a completed or accepted state - otherwise, return a null
	 * @param restApi
	 * @param story_id
	 * @return Story
	 * @throws IOException
	 */
	private static Story createStory(RallyRestApi restApi, String story_id) throws IOException {
		QueryRequest hir = new QueryRequest("hierarchicalrequirement");
		hir.setFetch(new Fetch("FormattedID", "ScheduleState"));
		QueryFilter filter = new QueryFilter("FormattedID", "=", story_id);
		hir.setQueryFilter(filter);
		QueryResponse queryResponse = restApi.query(hir);
		for (JsonElement result : queryResponse.getResults()) {
			HierarchyItem hi = gson.fromJson(result, HierarchyItem.class);
			System.out.println(hi.getFormattedID()+" - "+hi.getScheduleState());
			if (hi.getScheduleState() != null && (hi.getScheduleState().equalsIgnoreCase("Completed") || hi.getScheduleState().equalsIgnoreCase("Accepted"))) {
				Story newStory = new Story();
				newStory.setStory_id(story_id);
				return newStory;
			}
		}
		return null;
	}

	/**
	 * Calculates distribution of points based on hours worked by each team member
	 * @param currSprint
	 * @return Map<String,Story>
	 */
	private static Map<String, Story> distributePoints(Iteration currSprint) {
		Map<String,Story> storyMap = (currSprint.getStoryMap() == null) ? new HashMap<String,Story>() : currSprint.getStoryMap();
		for (String story_id : storyMap.keySet()) {
			Story currStory = storyMap.get(story_id);
			if (currStory.getAssigned_to().size() == 1) {
				//Only one person assigned to the story, therefore that person gets all the story points
				currStory.getAssigned_to().get(0).setUser_points(currStory.getStory_points());
			} else {
				//More than one person assigned to the story, therefore need to split based on the hours worked
				Double storyPoints = currStory.getStory_points();
				
				//Need to calculate the total hours to calculate what percentage of the points each person should receive
				Double totalHours = 0.0;
				for (AssignedTo currUser : currStory.getAssigned_to()) {
					totalHours += currUser.getUser_points();
				}
				
				//Can now calculate percentage of story points to assign to each person
				for (AssignedTo currUser : currStory.getAssigned_to()) {
					Double thisUserPoints = currUser.getUser_points()/totalHours*storyPoints;
					//Round it to the nearest thousandth
					thisUserPoints = ((int) ((thisUserPoints * 1000.0) + ((thisUserPoints < 0.0) ? -0.5 : 0.5))) / 1000.0;
					currUser.setUser_points(thisUserPoints);
				}
			}
		}
		return storyMap;
	}
	
}

File: fg_metrics/Rally/rally.js
// Usage: node ./rally.js [Teamname]
// Example: node ./rally.js CDMMS

const team = process.argv[2]
const config = require('./config/' + team + '.js')
const moment = require('moment')
const axios = require('axios')
const path = require('path')
const fs = require('fs')
const exportFile = moment().format('YYYYMMDD_HHmmss') + '_CTL_' + team + '.json'
const exec = require('child_process').exec

let startDate = moment().subtract(28, 'days').format()
let endDate = moment().add(14, 'days').format()

axios({
  method: 'get',
  url: `https://rally1.rallydev.com/slm/webservice/v2.0/iteration`,
  params: {
    apiKey: config.apiKey,
    fetch: 'Name,Notes,Owner,StartDate,EndDate',
    query: `(((Project.ObjectUUID = ${config.projectID}) and (StartDate >= ${startDate})) and (EndDate <= ${endDate}))`,
    pagesize: 200
  },
  headers: { 'zsessionid': config.apiKey }
}).then(async response => {
  if (response.data.QueryResult.Errors[0]) {
    console.error(response.data.QueryResult.Errors)
    process.exit()
  }

  let result = {
    team_id: config.team_id,
    system_id: config.projectID,
    sprint: []
  }

  for (let each in response.data.QueryResult.Results) {
    result.sprint.push(await buildResults(response.data.QueryResult.Results[each].Name))
  }

  // Write JSON results to result.json file
  fs.writeFile('C:/RallyStories/' + exportFile, JSON.stringify(result), function () {
    const child = exec('cmd /c C:/scripts/winscp.bat C:\\RallyStories\\' + exportFile, (error, stdout, stderr) => {
      console.log(`stdout: ${stdout}`)
      console.log(`stderr: ${stderr}`)

      if (error != null) {
        console.error(`exec error: ${error}`)
      }
    })
  })
}).catch(error => {
  console.error(error.stack)
})

function buildResults (sprintName) {
  return new Promise(async (resolve, reject) => {
    // Build query to get tasks assigned to members of the team
    let query = '('

    for (let each in config.userList) {
      if (each == 0) {
        query = query + `(Owner.Name = ${config.userList[each]})`
      } else {
        query = '(' + query + `(Owner.Name = ${config.userList[each]}))`
      }

      if (each < config.userList.length - 1) {
        query = query + ' or '
      }
    }

    query = query + ` and (Iteration.Name = "${sprintName}"))`

    // Get tasks assigned to users for specified sprint and initialize arrays
    let userResult = await getUsers(query)
    let hashArr = []
    let defects = []
    let userStories = []

    // Iterate over user tasks to create a hash and arrays of defects and user stories tasks are associated with
    for (let each in userResult) {
      if (!hashArr[userResult[each].WorkProduct.FormattedID]) {
        hashArr[userResult[each].WorkProduct.FormattedID] = []
        hashArr[userResult[each].WorkProduct.FormattedID].push({
          Actuals: userResult[each].Actuals,
          User: userResult[each].Owner['_refObjectName'].toLowerCase().replace(' ', '.') + '@centurylink.com'
        })
      } else {
        let flag = true

        for (let i in hashArr[userResult[each].WorkProduct.FormattedID]) {
          if (userResult[each].Owner['_refObjectName'].toLowerCase().replace(' ', '.') + '@centurylink.com' === hashArr[userResult[each].WorkProduct.FormattedID][i].User) {
            hashArr[userResult[each].WorkProduct.FormattedID][i].Actuals += userResult[each].Actuals
            flag = false
            break
          }
        }

        if (flag) {
          hashArr[userResult[each].WorkProduct.FormattedID].push({
            Actuals: userResult[each].Actuals,
            User: userResult[each].Owner['_refObjectName'].toLowerCase().replace(' ', '.') + '@centurylink.com'
          })
        }
      }

      if (userResult[each].WorkProduct.FormattedID.indexOf('DE') > -1 && defects.indexOf(userResult[each].WorkProduct.FormattedID) === -1) {
        defects.push(userResult[each].WorkProduct.FormattedID)
      } else if (userResult[each].WorkProduct.FormattedID.indexOf('US') > -1 && userStories.indexOf(userResult[each].WorkProduct.FormattedID) === -1) {
        userStories.push(userResult[each].WorkProduct.FormattedID)
      }
    }

    // Initialize variables for next section
    let startDate = ''
    let endDate = ''
    let stories = []

    // Pull the defects from Rally, get the start and end dates, and generate the results
    if (defects[0]) {
      var defectList = await getDefects(defects)
      startDate = moment(defectList[0].Iteration.StartDate).format('YYYY-MM-DD')
      endDate = moment(defectList[0].Iteration.EndDate).format('YYYY-MM-DD')
      stories = stories.concat(generateResults(defectList, hashArr))
    }

    // Pull the user stories from Rally, get the start and end dates, and generate the results
    if (userStories[0]) {
      var userStoryList = await getUserStories(userStories)
      startDate = moment(userStoryList[0].Iteration.StartDate).format('YYYY-MM-DD')
      endDate = moment(userStoryList[0].Iteration.EndDate).format('YYYY-MM-DD')
      stories = stories.concat(generateResults(userStoryList, hashArr))
    }

    resolve({
      sprint_id: sprintName,
      sprint_start_date: startDate,
      sprint_end_date: endDate,
      stories: stories
    })
  })
}

// REST call to get the tasks assigned to the members of the team
function getUsers (query) {
  return new Promise((resolve, reject) => {
    axios({
      method: 'get',
      url: `https://rally1.rallydev.com/slm/webservice/v2.0/task`,
      params: {
        apiKey: config.apiKey,
        fetch: 'WorkProduct,Owner,Name,State,Estimate,Actuals',
        query: query,
        pagesize: 200
      },
      headers: { 'zsessionid': config.apiKey }
    }).then(response => {
      if (response.data.QueryResult.Errors[0]) {
        reject(response.data.QueryResult.Errors)
        process.exit()
      }

      resolve(response.data.QueryResult.Results)
    }).catch(error => {
      reject(error.stack)
    })
  })
}

// REST call to get the defects that have tasks assigned to the team
function getDefects (defects) {
  return new Promise((resolve, reject) => {
    axios({
      method: 'get',
      url: `https://rally1.rallydev.com/slm/webservice/v2.0/defects`,
      params: {
        apiKey: config.apiKey,
        fetch: 'FormattedID,ScheduleState,PlanEstimate,Iteration,StartDate,EndDate',
        query: generateQuery(defects),
        pagesize: 200
      },
      headers: { 'zsessionid': config.apiKey }
    }).then(response => {
      if (response.data.QueryResult.Errors[0]) {
        reject(response.data.QueryResult.Errors)
        process.exit()
      }

      resolve(response.data.QueryResult.Results)
    }).catch(error => {
      reject(error.stack)
    })
  })
}

// REST call to get the user stories that have tasks assigned to the team
function getUserStories (userStories) {
  return new Promise((resolve, reject) => {
    axios({
      method: 'get',
      url: `https://rally1.rallydev.com/slm/webservice/v2.0/hierarchicalrequirement`,
      params: {
        apiKey: config.apiKey,
        fetch: 'FormattedID,ScheduleState,PlanEstimate,Iteration,StartDate,EndDate',
        query: generateQuery(userStories),
        pagesize: 200
      },
      headers: { 'zsessionid': config.apiKey }
    }).then(response => {
      if (response.data.QueryResult.Errors[0]) {
        reject(response.data.QueryResult.Errors)
        process.exit()
      }

      resolve(response.data.QueryResult.Results)
    }).catch(error => {
      reject(error.stack)
    })
  })
}

// Generate the queries for defects and user stories
function generateQuery (arr) {
  query = '('

  for (let each in arr) {
    if (each == 0) {
      query = query + `(FormattedID = ${arr[each]})`
    } else {
      query = '(' + query + `(FormattedID = ${arr[each]}))`
    }

    if (each < arr.length - 1) {
      query = query + ' or '
    }
  }

  query = query + ' and ((ScheduleState = Completed) or (ScheduleState = Accepted)))'

  return query
}

// Generate and distribute points to the team
function generateResults (arr, hashArr) {
  let stories = []

  for (let i in arr) {
    let totalHours = 0
    let assignedTo = []

    for (let j in hashArr[arr[i].FormattedID]) {
      totalHours += hashArr[arr[i].FormattedID][j].Actuals
    }

    for (let j in hashArr[arr[i].FormattedID]) {
      let points = +(((hashArr[arr[i].FormattedID][j].Actuals / totalHours) * arr[i].PlanEstimate).toFixed(3)) || 0
      assignedTo.push({ user: hashArr[arr[i].FormattedID][j].User, user_points: points })
    }

    stories.push({
      story_id: arr[i].FormattedID,
      story_points: arr[i].PlanEstimate,
      assigned_to: assignedTo
    })
  }

  return stories
}
